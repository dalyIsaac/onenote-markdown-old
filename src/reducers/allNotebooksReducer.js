import { PUT_ALL_NOTEBOOKS, CLEAR_ALL_NOTEBOOKS } from "./../actionTypes";

export default function notebooksReducer(state = [], action) {
  switch (action.type) {
    case CLEAR_ALL_NOTEBOOKS:
      return [];
    case PUT_ALL_NOTEBOOKS:
      return [
        ...state,
        {
          userId: action.userId,
          displayableId: action.displayableId,
          notebooks: action.notebooks
        }
      ];
    default:
      return state;
  }
}
