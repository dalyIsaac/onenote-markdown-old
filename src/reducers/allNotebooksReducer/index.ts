import notebooksReducer from "./allNotebooksReducer";
export default notebooksReducer;
