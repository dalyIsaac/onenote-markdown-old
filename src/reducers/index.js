import { combineReducers } from 'redux';
import userReducer from './userReducer';

export default combineReducers({
    users: userReducer
});
