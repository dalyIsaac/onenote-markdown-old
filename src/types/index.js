// import { Notebook } from '@microsoft/microsoft-graph-types/microsoft-graph'; // eslint-disable-line

/**
 * Custom class which extends MSAL's User class.  
 * This class is not deflated in order to reduce boilerplate. 
 * @export
 * @class UserData
 */
export class UserData {
    constructor(msal, photo = "", acquireTokenError = null) {
        this.displayableId = undefined;
        Object.assign(this, msal);
        this.photo = photo;
        this.acquireTokenError = acquireTokenError;
    }
}

/**
 * Used in the DetailsList inside NotebookPicker
 * @export
 * @class NotebookRow
 */
export class NotebookRow {
    /**
     * Creates an instance of NotebookRow.
     * @param {Object} notebook 
     * @param {string} userId
     * @memberof NotebookRow
     */
    constructor(notebook, userId, displayableId) {
        this.fileName = notebook.displayName;
        this.lastModifiedDateTime = new Date(notebook.lastModifiedDateTime);
        this.userDisplayableId = displayableId;
        this.userId = userId;
        this.notebook = notebook;
    }
}

/**
 * Deflated notebook from Microsoft Graph with some custom attributes
 * @class Notebook
 */
export class Notebook {
    /**
     * @param {Object} notebook JSON response from the Microsoft Graph for a notebook
     * @param {string} userId
     */
    constructor(notebook, userId = undefined) {
        this.id = undefined; // the following explicitly defined properties are here purely for VSCode
        this.sectionGroups = [];
        this.sections = [];
        if (userId !== undefined) {
            this.userId = userId;
        }
        deflateObject(this, notebook);
    }
}

/**
 * Deflated section group from Microsoft Graph with some custom attributes
 * @export
 * @class SectionGroup
 */
export class SectionGroup {
    constructor(sectionGroup, userId = undefined) {
        this.id = undefined; // the following explicitly defined properties are here purely for VSCode
        this.sectionGroups = [];
        this.sections = [];
        this.isExpanded = false;
        if (userId !== undefined) {
            this.userId = userId;
        }
        deflateObject(this, sectionGroup);
    }
}

/**
 * Deflated section group from Microsoft Graph with some custom attributes
 * @export
 * @class Section
 */
export class Section {

    /**
     * @param {Object} section Result from the Microsoft Graph query
     * @param {*} pages Result from the Microsoft Graph query on the /pages method
     * @param {*} userId 
     */
    constructor(section, pages, userId = undefined) {
        this.id = undefined; // the following explicitly defined properties are here purely for VSCode
        this.pages = getIds(pages.value);
        if (userId !== undefined) {
            this.userId = userId;
        }
        deflateObject(this, section);
    }
}

export class Page {
    /**
     * @param {Object} page Result from the Microsoft Graph query
     * @param {string} content Result from the Microsoft Graph query on the /content method
     * @param {string} userId 
     */
    constructor(page, content, userId = undefined) {
        this.id = undefined; // the following explicitly defined properties are here purely for VSCode
        this.content = content;
        if (userId !== undefined) {
            this.userId = userId;
        }
        deflateObject(this, page);
    }
}

/**
 * Deflates an object.  
 * For example: `{ "hello": { "world": "earth" } }` goes to `{ "hello.world": "earth" }`  
 * Note: a key which begins with `@` will not be deflated.
 * @param {Object} target 
 * @param {Object} source 
 * @param {string} [parentName='']
 */
export function deflateObject(target, source, parentName = '') {
    for (const key in source) {
        const prop = source[key];
        if (Array.isArray(prop)) {
            if (parentName !== '') {
                target[`${parentName}.${key}`] = getIds(prop);
            } else {
                target[key] = getIds(prop);
            }
        } else if (typeof prop === 'object') {
            if (parentName !== '') {
                deflateObject(target, prop, `${parentName}.${key}`);
            } else {
                deflateObject(target, prop, `${key}`);
            }
        } else {
            if (parentName !== '') {
                target[`${parentName}.${key}`] = prop;
            } else {
                target[key] = prop;
            }
        }
    }
}

/**
 * Gets the IDs from an array of objects
 * @param {array} element
 * @returns {array} idList List of IDs
 */
function getIds(element) {
    let idList = [];
    element.forEach(element => {
        idList.push(element.id);
    });
    return idList;
}

/**
 * Inflates an object.
 * For example: `{ "hello.world": "earth" }` goes to `{ "hello": { "world": "earth" } }` 
 * @param {any} target 
 * @param {any} source 
 * @param {any} key 
 */
function inflateObject(target, source, key) {
    if (typeof source === 'object') {
        for (const key in source) {
            _inflateObjectHelper(target, source[key], key)
        }
    } else {
        _inflateObjectHelper(target, source, key)
    }
}

function _inflateObjectHelper(target, source, key) {
    const dotIndex = key.includes('@odata') ? -1 : key.indexOf('.');

    if (dotIndex === -1) {
        target[key] = source;
    } else {
        const parentKey = key.slice(0, dotIndex);
        const childKey = key.slice(dotIndex + 1);
        if (target[parentKey] === undefined) {
            target[parentKey] = {};
        }
        inflateObject(target[parentKey], source, childKey)
    }
}