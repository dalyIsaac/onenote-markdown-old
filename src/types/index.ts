/**
 * Deflates an object.  
 * For example: `{ "hello": { "world": "earth" } }` goes to `{ "hello.world": "earth" }`  
 * Note: a key which begins with `@` will not be deflated.
 */
export function deflateObject(target: object, source: object, parentName: string = '') {
    for (const key in source) {
        if (source.hasOwnProperty(key)) {
            const prop = source[key];
            if (Array.isArray(prop)) {
                if (parentName !== '') {
                    target[`${parentName}.${key}`] = getIds(prop);
                } else {
                    target[key] = getIds(prop);
                }
            } else if (typeof prop === 'object') {
                if (parentName !== '') {
                    deflateObject(target, prop, `${parentName}.${key}`);
                } else {
                    deflateObject(target, prop, `${key}`);
                }
            } else {
                if (parentName !== '') {
                    target[`${parentName}.${key}`] = prop;
                } else {
                    target[key] = prop;
                }
            }
        }
    }
}

/**
 * Gets the IDs from an array of objects
 * @param {array} element
 * @returns {array} idList List of IDs
 */
function getIds(element: any[]): string[] {
    /* tslint:disable-next-line */
    let idList: string[] = [];
    element.forEach((el: any) => {
        idList.push(el.id);
    });
    return idList;
}

/**
 * Inflates an object.
 * For example: `{ "hello.world": "earth" }` goes to `{ "hello": { "world": "earth" } }` 
 */
function inflateObject(target: any, source: any, key: any) {
    if (typeof source === 'object') {
        for (const index in source) {
            if (source.hasOwnProperty(key)) {
                _inflateObjectHelper(target, source[index], key)
            }
        }
    } else {
        _inflateObjectHelper(target, source, key)
    }
}

function _inflateObjectHelper(target: any, source: any, key: any) {
    const dotIndex = key.includes('@odata') ? -1 : key.indexOf('.');

    if (dotIndex === -1) {
        target[key] = source;
    } else {
        const parentKey = key.slice(0, dotIndex);
        const childKey = key.slice(dotIndex + 1);
        if (target[parentKey] === undefined) {
            target[parentKey] = {};
        }
        inflateObject(target[parentKey], source, childKey)
    }
}