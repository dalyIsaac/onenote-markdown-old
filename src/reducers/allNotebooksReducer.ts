import { IUserNotebooks } from "src/reducers";
import { IAction } from "../actions";
import { IPutAllNotebooks } from "../actions/getNotebooks";
import { CLEAR_ALL_NOTEBOOKS, PUT_ALL_NOTEBOOKS } from "./../actionTypes";

type actionType = IPutAllNotebooks & IAction;

export default function notebooksReducer(
  state: IUserNotebooks[] = [],
  action: actionType
) {
  switch (action.type) {
    case CLEAR_ALL_NOTEBOOKS:
      return [];
    case PUT_ALL_NOTEBOOKS:
      return [
        ...state,
        {
          displayableId: action.displayableId,
          notebooks: action.notebooks,
          userId: action.userId
        }
      ];
    default:
      return state;
  }
}
