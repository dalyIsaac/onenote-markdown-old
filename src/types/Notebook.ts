import { Notebook as INotebook } from "@microsoft/microsoft-graph-types";
import { deflateObject } from "src/types";
import { OneNoteBase } from "./OneNoteBase";
import { SectionGroup } from "./SectionGroup";

/**
 * Deflated notebook from Microsoft Graph with some custom attributes
 * @class Notebook
 */
export class Notebook extends OneNoteBase {
  public sectionGroups: SectionGroup[];
  public sections: string[];
  public userId?: string;

  /**
   * @param {Object} notebook JSON response from the Microsoft Graph for a notebook
   * @param {string} userId
   */
  constructor(notebook: INotebook, userId?: string) {
    super();
    this.userId = userId;
    deflateObject(this, notebook);
  }
}
