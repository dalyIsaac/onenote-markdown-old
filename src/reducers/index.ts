import { Notebook as Graph_Notebook } from "@microsoft/microsoft-graph-types";
import { RouterState } from "connected-react-router";
import { combineReducers } from "redux";
import { Notebook } from "src/types/Notebook";
import { Page } from "src/types/Page";
import { Section } from "src/types/Section";
import { SectionGroup } from "src/types/SectionGroup";
import { UserData } from "src/types/UserData";
import allNotebooksReducer from "./allNotebooksReducer";
import notebookOrderReducer from "./notebookOrderReducer";
import onenoteReducer from "./onenoteReducer";
import selectedNavReducer from "./selectedNavReducer";
import totalNotebookLengthReducer from "./totalNotebookLengthReducer";
import userReducer from "./userReducer";

export default combineReducers({
  allNotebooks: allNotebooksReducer, // used for opening notebooks
  notebookOrder: notebookOrderReducer,
  onenote: onenoteReducer,
  selectedNav: selectedNavReducer,
  totalNotebookLength: totalNotebookLengthReducer,
  users: userReducer
});

export interface IState {
  allNotebooks: IStateUserNotebooks[];
  notebookOrder: string[];
  onenote: IStateOneNote[];
  router: RouterState;
  selectedNav: string[];
  totalNotebookLength: number;
  users: IStateUsers;
}

export interface IStateUserNotebooks {
  userId: string;
  displayableId: string;
  /**
   * This is not `Notebook` as defined in this project, but `Graph_Notebook` as an alias for `Notebook` from `@microsoft/microsoft-graph-types`.
   * This is because the `allNotebooks` slice of the state only stores the raw notebooks from Microsoft Graph.
   */
  notebooks: Graph_Notebook[];
}

export interface IStateOneNote {
  [key: string]: Notebook | SectionGroup | Section | Page;
}

export interface IStateUsers {
  [key: string]: UserData;
}
