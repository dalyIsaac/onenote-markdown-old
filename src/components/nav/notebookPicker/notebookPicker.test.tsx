import * as enzyme from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import {
  DefaultButton,
  DetailsList,
  IColumn,
  IObjectWithKey,
  SelectionMode
} from "office-ui-fabric-react";
import {
  ISearchBoxProps,
  MarqueeSelection,
  SearchBox,
  setIconOptions
} from "office-ui-fabric-react";
import * as React from "react";
import * as renderer from "react-test-renderer";
import { IStateOneNote, IStateUserNotebooks } from "../../../reducers";
import {
  graphNotebookInstance,
  graphNotebookInstance1,
  graphNotebookInstance2,
  notebook,
  notebook1,
  notebook2,
  userId,
  userId1
} from "../../../testObjects";
import { Notebook } from "../../../types/Notebook";
import NotebookPicker from "./index";

// Suppress icon warnings.
setIconOptions({
  disableWarnings: true
});

enzyme.configure({ adapter: new Adapter() });

function setUp() {
  const allNotebooks: IStateUserNotebooks[] = [
    {
      displayableId: "genericstring",
      notebooks: [graphNotebookInstance, graphNotebookInstance2],
      userId
    },
    {
      displayableId: "genericstring1",
      notebooks: [graphNotebookInstance1],
      userId: userId1
    }
  ];
  const openedNotebooks: IStateOneNote = {};
  openedNotebooks[notebook.id] = notebook;
  const props = {
    allNotebooks,
    closeModal: jest.fn(),
    openNotebooks: jest.fn((notebookList: Notebook[]) => undefined),
    openedNotebooks
  };

  const wrapper = enzyme.mount(<NotebookPicker {...props} />);

  return {
    props,
    wrapper
  };
}

function columnTestSetUp(): IColumn[] {
  const { wrapper } = setUp();
  return (wrapper.instance() as NotebookPicker).state.columns;
}

describe("Component: notebookPicker", () => {
  test("It should render self and subcomponents", () => {
    const { wrapper } = setUp();

    // notebookPickerWrapper
    expect(
      wrapper
        .find("div")
        .first()
        .hasClass("notebookPickerWrapper")
    ).toBe(true);

    // SearchBox
    const searchBox = wrapper.find(SearchBox);
    expect(searchBox.parent().is("div.notebookPickerWrapper")).toEqual(true);
    const searchBoxProps: ISearchBoxProps = searchBox.props();
    expect(searchBoxProps.className).toBe("filterDiv");
    expect(searchBoxProps.placeholder).toBe("Filter");

    // detailsListDiv
    expect(
      wrapper
        .find("div.detailsListDiv")
        .parent()
        .is("div.notebookPickerWrapper")
    ).toEqual(true);

    // MarqueeSelection
    const marqueeSelection = wrapper.find(MarqueeSelection);
    expect(marqueeSelection.parent().is("div.detailsListDiv")).toEqual(true);
    expect(
      marqueeSelection
        .children()
        .find(DetailsList)
        .exists()
    ).toBe(true);

    // DetailsList
    const detailsList = wrapper.find(DetailsList);
    expect(detailsList.props().items).toEqual([notebook2, notebook1]);
    expect(detailsList.props().isHeaderVisible).toBe(true);
    expect(detailsList.props().selectionMode).toBe(SelectionMode.multiple);

    // DefaultButton
    const defaultButton = wrapper.find(DefaultButton);
    expect(defaultButton.parent().is("div.notebookPickerWrapper")).toBe(true);
    expect(defaultButton.props().className).toBe("footerDiv");
    expect(defaultButton.props().primary).toBe(true);
    expect(defaultButton.props().secondaryText).toBe(
      "Starts opening the selected notebooks"
    );
    expect(defaultButton.props().text).toBe("Open notebooks");
  });

  describe("It should sort the notebooks by the column clicked", () => {
    const testColumn = (
      column: IColumn,
      test1Results: Notebook[],
      test2Results: Notebook[],
      test3Results: Notebook[]
    ) => {
      const { wrapper } = setUp();
      const notebookPicker = wrapper.instance() as NotebookPicker;
      const detailsList = wrapper.find(DetailsList).instance() as DetailsList;

      notebookPicker.onColumnClick({}, column);
      expect(notebookPicker.state.notebooks).toEqual(test1Results);
      expect(detailsList.props.items).toEqual(test1Results);

      notebookPicker.onColumnClick(
        {},
        { ...column, isSortedDescending: !column.isSortedDescending }
      );
      expect(notebookPicker.state.notebooks).toEqual(test2Results);
      expect(detailsList.props.items).toEqual(test2Results);

      notebookPicker.setState({ notebooks: [notebook1, notebook2] });
      notebookPicker.onColumnClick({}, { ...column, isSortedDescending: true });
      expect(notebookPicker.state.notebooks).toEqual(test3Results);
      expect(detailsList.props.items).toEqual(test3Results);
    };

    const displayNameColumn: IColumn = {
      data: "string",
      fieldName: "displayName",
      isPadded: true,
      isResizable: true,
      isRowHeader: true,
      isSortedDescending: true,
      key: "displayNameColumn",
      maxWidth: 350,
      minWidth: 210,
      name: "Name"
    };

    test("It should sort the notebooks by the displayName column", () => {
      testColumn(
        displayNameColumn,
        [notebook1, notebook2],
        [notebook2, notebook1],
        [notebook1, notebook2]
      );
    });

    const lastModifiedDateTimeColumn: IColumn = {
      fieldName: "lastModifiedDateTime",
      isPadded: true,
      isResizable: true,
      isRowHeader: true,
      isSorted: true,
      isSortedDescending: true,
      key: "lastModifiedDateTimeColumn",
      maxWidth: 350,
      minWidth: 210,
      name: "Last Modified Date Time"
    };

    test("It should sort the notebooks by the lastModifiedDateTime column", () => {
      testColumn(
        lastModifiedDateTimeColumn,
        [notebook1, notebook2],
        [notebook2, notebook1],
        [notebook1, notebook2]
      );
    });

    const userDisplayableIdColumn: IColumn = {
      data: "string",
      fieldName: "userDisplayableId",
      isPadded: true,
      isResizable: true,
      isRowHeader: true,
      isSortedDescending: false,
      key: "userDisplayableIdColumn",
      maxWidth: 350,
      minWidth: 210,
      name: "Account"
    };

    test("It should sort the notebooks by the userDisplayableId column", () => {
      testColumn(
        userDisplayableIdColumn,
        [notebook1, notebook2],
        [notebook2, notebook1],
        [notebook2, notebook1]
      );
    });

    const iconColumn: IColumn = {
      fieldName: "icon",
      iconName: "Page",
      isIconOnly: true,
      key: "iconColumn",
      maxWidth: 16,
      minWidth: 16,
      name: "Icons"
    };

    test("It should attempt to sort the notebooks by the icon column (in reality, no sorting actually happens)", () => {
      testColumn(
        iconColumn,
        [notebook2, notebook1],
        [notebook2, notebook1],
        [notebook1, notebook2]
      );
    });
  });

  test("Each column should have this.onColumnClick bound to onColumnClick", () => {
    const { wrapper } = setUp();
    const notebookPicker = wrapper.instance() as NotebookPicker;
    const columns = notebookPicker.state.columns;
    for (let i = 1; i < columns.length; i++) {
      const column = columns[i];
      expect(column.onColumnClick).toBe(notebookPicker.onColumnClick);
    }
  });

  test("It should filter based on the search box input", () => {
    const { wrapper } = setUp();
    const notebookPicker = wrapper.instance() as NotebookPicker;
    notebookPicker.updateFilter("id1");
    expect(notebookPicker.state.notebooks).toEqual([notebook1]);
    notebookPicker.updateFilter("");
    expect(notebookPicker.state.notebooks).toEqual([notebook2, notebook1]);
    notebookPicker.updateFilter("id2");
    expect(notebookPicker.state.notebooks).toEqual([notebook2]);
    notebookPicker.updateFilter("notebook");
    expect(notebookPicker.state.notebooks).toEqual([notebook2, notebook1]);
  });

  test("Checks detailsList can select multiple objects", () => {
    const { wrapper } = setUp();
    const detailsList = wrapper.find(DetailsList).instance() as DetailsList;
    expect(detailsList.props.selectionMode).toBe(SelectionMode.multiple);
  });

  test("Checks that openNotebooks calls the action creator to open the selected notebooks", () => {
    const { wrapper } = setUp();
    const notebookPicker = wrapper.instance() as NotebookPicker;
    notebookPicker.openNotebooks();
    expect(
      (notebookPicker.props.openNotebooks as jest.Mock).mock.calls.length
    ).toBe(0);

    notebookPicker.selection.getSelectedCount = jest.fn().mockReturnValue(1);
    notebookPicker.selection.getSelection = jest
      .fn()
      .mockReturnValue([notebook1] as IObjectWithKey[]);
    notebookPicker.openNotebooks();
    expect(
      (notebookPicker.props.openNotebooks as jest.Mock).mock.calls.length
    ).toBe(1);
    expect(
      (notebookPicker.props.openNotebooks as jest.Mock).mock.calls[0]
    ).toEqual([[notebook1]]);
  });

  test("Checks that the images column renders an image", () => {
    const column = columnTestSetUp()[0];
    if (column.onRender !== undefined) {
      const renderedColumn = enzyme.shallow(column.onRender());
      expect(renderedColumn.find("img").exists()).toBe(true);

      const component = renderer.create(column.onRender());
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    }
  });

  test("Checks that the last date modified is correct", () => {
    const column = columnTestSetUp()[3];
    if (column.onRender !== undefined) {
      const component = renderer.create(column.onRender(notebook));
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();

      const notebookWithNoDate = notebook1;
      notebookWithNoDate.lastModifiedDateTime = undefined;
      const component1 = renderer.create(column.onRender(notebookWithNoDate));
      const tree1 = component1.toJSON();
      expect(tree1).toMatchSnapshot();
    }
  });
});
