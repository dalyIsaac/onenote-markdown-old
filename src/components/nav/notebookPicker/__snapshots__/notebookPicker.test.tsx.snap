// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: NotebookPicker Checks that the images column renders an image 1`] = `
<img
  alt="OneNote icon"
  src="https://static2.sharepointonline.com/files/fabric/assets/brand-icons/document/svg/one_16x1.svg"
/>
`;

exports[`Component: NotebookPicker Checks that the last date modified is correct 1`] = `
<div>
  Invalid Date
</div>
`;

exports[`Component: NotebookPicker Checks that the last date modified is correct 2`] = `
<div>
  N/A
</div>
`;
